docs/architecture-decision-records/2023-03-integration-with-station.md

```diff
@@ -0,0 +1,232 @@
+# Zinnia in Filecoin Station
+
+<!--
+This is a minimal template. Feel free to add more sections as needed.
+
+Please review also the Design Doc template and add any relevant sections to your ADR:
+https://www.notion.so/pl-strflt/Writing-a-Design-Doc-aa6034be43c2434ba88a2fd844516e94
+-->
+
+> Status: ACCEPTED
+
+<!--
+PROPOSED, ACCEPTED, REJECTED, DEPRECATED, SUPERSEDED BY {link-to-ADR}
+-->
+
+## Context
+
+Zinnia is a runtime for Filecoin Station modules. It's time to integrate Zinnia into the Station to
+enable the deployment of the first modules.
+
+Zinnia is designed to run multiple modules inside the same process, using V8 Isolates for keeping
+modules cleanly separated and sandboxed. The integration with libp2p (and later IPFS) is designed to
+allow multiple modules to share the same underlying set of network connections and block storage.
+
+Zinnia has two primary modes of operation:
+
+1. A developer tool for building a single module. We want to optimize for ease of use and a fast
+   feedback loop.
+
+   In this mode, Zinnia is typically started and stopped frequently, running for short periods at a
+   time, executing a single module only.
+
+   As a CLI tool, it reads the configuration from a config file (either user-level or
+   project-specific) or CLI arguments. There should be reasonable defaults tailored to developers
+   building a module - e.g. the state files should be stored in the project's working directory.
+
+2. A deployment platform, running modules inside Filecoin Station instances operated by
+   non-technical users. We want to optimize for a smooth user experience for Station operators
+   first, and easy troubleshooting of operations by module developers second.
+
+   In this mode, Zinnia is a long-running process executing multiple modules. It needs to support
+   reloading individual modules when a new version of a module is deployed. It also needs to report
+   various information back to the Station, e.g. activity log and the number of jobs completed.
+
+   As a service, it should read the configuration from environment variables (see
+   [The Twelve-Factor App](https://12factor.net/config)) and allow embedders to improve security
+   e.g. by keeping modules in a different place from the runtime state.
+
+### `zinniad`
+
+For the reasons above, I propose to build a new binary called `zinniad` that will execute Zinnia
+modules inside the Station.
+
+_(This is a strawman proposal, subject to changes during implementation.)_
+
+```bash
+# Address of Station's built-in wallet
+FIL_WALLET_ADDRESS=f1etc \
+# Where to keep state files
+STATE_ROOT=$HOME/Library/Caches/Filecoin\ Station/zinnia \
```

Comment:
Great question. The major difference between `XDG_STATE_HOME` and `STATE_ROOT` is that `XDG_STATE_HOME` provides a system (or user) wide directory, we need to append a zinnia-specific segment to that path to obtain `STATE_ROOT`.

I prefer to give the user full control over the location by providing them `STATE_ROOT` config option.

This becomes somewhat relevant when Zinnia is running inside the Station.
- If we use `XDG_STATE_HOME`, then we will keep the state in `XDG_STATE_HOME/zinnia`, a different place from where Station keeps its files. We could make this path Station specific, but that feels hacky to me and incorrect in the situation when Zinnia runs outside of the Station

However, I think it's a good idea to make `STATE_ROOT` an optional configuration option and use `XDG_STATE_HOME/zinnia` as the default value.

WDYT?

Reply:
Edit: let's use `XDG_STATE_HOME/zinniad` as the default value to avoid potential conflict with state files used by `zinnia` (the CLI for development)

Reply:
I have updated the proposal - see [f41b8df](https://github.com/filecoin-station/zinnia/pull/121/commits/f41b8df6fcf308c199cfd43cc2b9b9419e0ade2a)

Reply:
I also find `XDG_STATE_HOME` inflexible because the app needs to pick where to store its file inside that directory. If you wanted to tell `zinniad` to store files in a specific directory, you could pass `XDG_STATE_HOME=$XDG_STATE_HOME/zinnia2`, however then the resulting path would be an ugly `$XDG_STATE_HOME/zinnia2/zinnia`.

> If we use XDG_STATE_HOME, then we will keep the state in XDG_STATE_HOME/zinnia, a different place from where Station keeps its files. We could make this path Station specific, but that feels hacky to me and incorrect in the situation when Zinnia runs outside of the Station

This is a good question. If Station launches Zinnia, should Zinnia store its files in a Station directory, or in its own default directory? For now I think until this becomes a problem we can have independent directories and don't need to put Zinnia stuff inside Station folders.

Reply:
> This is a good question. If Station launches Zinnia, should Zinnia store its files in a Station directory, or in its own default directory? For now I think until this becomes a problem we can have independent directories and don't need to put Zinnia stuff inside Station folders.

I prefer to keep Zinnia files inside Station folders. 

As a user running the Station on my desktop, when inspecting files created by various apps, I prefer apps to have all files in a single directory with the app name (e.g. `$HOME/Library/Caches/Filecoin\ Station`). If we use different locations for different parts of the Station, then it's not clear which directories like `zinnia` and `saturn-l2` belong to the Station and which are created by other apps.

Reply:
Ok gotcha! What do you think about this pattern, where we use `XDG_STATE_HOME` but don't need `STATE_ROOT`:

The app that hosts another app (here Station hosting Zinnia) will pass on a `XDG_STATE_HOME` like this:

```
XDG_STATE_HOME=$XDG_STATE_HOME/filecoin-station/dependencies zinniad
```

That way Zinnia's state will live in

```
$XDG_STATE_HOME/filecoin-station/dependencies/zinniad
```

Reply:
I see where you are going here and I am not entirely opposed.

However, there is another catch: `XDG_STATE_HOME` is Linux specific. Windows uses something like `%LOCALAPPDATA%`, macOS something else.

We already provide `ROOT_DIR` for modules, should I perhaps rename `STATE_ROOT` to `ROOT_DIR` for consistency?

https://github.com/filecoin-station/desktop/blob/596b16704347c5c6e09c621f475cd20291938b77/docs/MODULE_INTERFACE.md?plain=1#L27-L29
> `ROOT_DIR` The long-lived working directory on disk. The module must store all of its files inside (subdirectories of) this directory. The directory isn't expected to be backed up or shared across machines in any way.

Reply:
Good point. If it makes everything easier, let's just use one variable that is consistent across OS, but with per-OS defaults. +1 to `ROOT_DIR` for consistency!

Reply:
See 758c6b0